### PR TITLE
Add test fixtures using rector

### PR DESCRIPTION
**Pull request by jtojnar**
Use `composer rector` (or `composer rector -- --dry-run`) to run the automated refactoring.

### CODE REVIEW

**Comment by j0k3r** [comment]
Maybe you can separate the code improvement (`fn`, `??`, etc.) in a separate PR to reduce the noise?

**Comment by jtojnar** [comment]
I can separate them out but then CI will not pass.

**Comment by jtojnar** [comment]
Or I can move the ci commit to the new PR too.

**Comment by jtojnar** [comment]
Removed the ci part for now.

Still need to replace `file_get_contents` with actual `fetchContent` and capture the resources.

**Comment by jtojnar** [comment]
I got nerd sniped quite hard by this but I had this on to-do list for few other projects. Now running `composer rector` should automatically add fixtures for `fetchContent` calls on `Graby` constructed without a second argument.

**Review by j0k3r — COMMENTED** [review]
I don't get everything about Rector (never used it).
What I understand if I add a new functional test is I'll need to run it once and then the fixture html will be generated and I'll need to update the functional test to use the fixture.
Am I right ?
If so, how do you catch headers you added in the test file?

**Comment by jtojnar** [comment]
Rector is a tool for automated code transformation. The transformation rule runs Graby and extract headers using `RecordingHttpClient` will also update the functional tests – the last commit is fully produced by the transformation rule without any manual changes.

**Comment by jtojnar** [comment]
If you check out the PR, you can try it yourself by removing the last commit and running `composer rector`. (comment out [this line](https://github.com/jtojnar/graby/blob/a2d0269fd3060f84978b60fadc0b85ec5dfa9776/maintenance/rector.config.php#L12) if you do not want the code modernization).

File: tests/GrabyFunctionalTest.php [hunk 1]
```diff
@@ -312,14 +328,17 @@ public function testMultipage(): void
 
     public function testCookie(): void
     {
+        $httpMockClient = new HttpMockClient();
+        $httpMockClient->addResponse(new Response(301, ['Server' => ['AkamaiGHost'], 'Content-Length' => ['0'], 'Location' => ['https://www.npr.org/sections/parallels/2017/05/19/529148729/michael-flynns-contradictory-line-on-russia'], 'Date' => ['Sat, 08 Jan 2022 05:02:29 GMT'], 'Connection' => ['keep-alive']], (string) file_get_contents(__DIR__ . '/fixtures/content/http___www.npr.org_sections_parallels_2017_05_19_529148729_michael-flynns-contradictory-line-on-russia.html')));
```
Comment (jtojnar) [inline tests/GrabyFunctionalTest.php:332]: Perhaps we should update the link to remove the redirect.

File: tests/GrabyFunctionalTest.php [hunk 2]
```diff
@@ -312,14 +328,17 @@ public function testMultipage(): void
...
+        $httpMockClient->addResponse(new Response(200, ['Content-Type' => ['text/html;charset=UTF-8'], 'X-Cache-NPR' => ['MISS'], 'Access-Control-Allow-Origin' => ['http://www.google.co.uk'], 'Access-Control-Allow-Credentials' => ['true'], 'X-NPR-Trace-ID' => ['renIaLEEV1Q'], 'X-Content-Type-Options' => ['nosniff'], 'X-XSS-Protection' => ['1; mode=block'], 'X-Served-By' => ['pod-www-render-nginx-776fbf4c8c-7sjbx'], 'Referrer-Policy' => ['no-referrer-when-downgrade'], 'Strict-Transport-Security' => ['max-age=15724800; includeSubDomains'], 'Cache-Control' => ['no-cache'], 'Expires' => ['Sat, 08 Jan 2022 05:02:30 GMT'], 'Date' => ['Sat, 08 Jan 2022 05:02:30 GMT'], 'Transfer-Encoding' => ['chunked'], 'Connection' => ['keep-alive', 'Transfer-Encoding']], (string) file_get_contents(__DIR__ . '/fixtures/content/http___www.npr.org_sections_parallels_2017_05_19_529148729_michael-flynns-contradictory-line-on-russia')));
```
Comment (jtojnar) [inline tests/GrabyFunctionalTest.php:333]: I guess this will not actually do anything if we want to send cookie.

Comment (j0k3r) [reply tests/GrabyFunctionalTest.php:333]: You're right, could you add an exclusion rule for Rector to not auto mock that request?

Comment (jtojnar) [reply tests/GrabyFunctionalTest.php:333]: There are three options for this:

- Have blacklist of function names
- Use some in code marker (e.g. a comment like `Rector: do not add mocking`
- Pass `null` as second argument (with a comment for humans that we do not want mocking) – this one works now.

What would you prefer.

Comment (j0k3r) [reply tests/GrabyFunctionalTest.php:333]: The code marker is an annotation to the function to blacklist?

Comment (j0k3r) [reply tests/GrabyFunctionalTest.php:333]: I'm ok with that

Comment (jtojnar) [reply tests/GrabyFunctionalTest.php]: It is now skipped.

**Comment by j0k3r** [comment]
Can you rebase against the master?
And it'll be good to merge

**Comment by j0k3r** [comment]
Good job @jtojnar 👊

**Pull request merged**